### PR TITLE
Let sensors also export state interfaces of joints

### PR DESCRIPTION
**Pull request by christophfroehlich**
- Let sensor export also state interfaces of joints
- Fix *returned no state interfaces* warning

needed for https://github.com/ros-controls/ros2_control_demos/pull/653

### CODE REVIEW

**Review by saikishor — APPROVED** [review]
In general, LGTM

File: hardware_interface/src/resource_manager.cpp [hunk 1]
```diff
@@ -611,11 +611,13 @@ class ResourceStorage
...
-      "Importing state interfaces for the hardware '%s' returned no state interfaces.",
-      hardware.get_name().c_str());
-
+    if (interface_names.empty())
+    {
+      RCLCPP_WARN(
+        get_logger(),
+        "Importing state interfaces for the hardware '%s' returned no state interfaces.",
+        hardware.get_name().c_str());
+    }
```
[the remark this answers is not in the document]
Comment (christophfroehlich) [reply hardware_interface/src/resource_manager.cpp:620]: I always learn something new

**Review by saikishor — DISMISSED** [review]
Thanks for fixing it

**Review by bmagyar — APPROVED** [review]

File: hardware_interface/test/test_component_interfaces.cpp [hunk 2]
```diff
@@ -948,29 +1007,86 @@ TEST(TestComponentInterfaces, dummy_sensor_default)
...
+  rclcpp::Node::SharedPtr node = std::make_shared<rclcpp::Node>("test_system_components");
+  auto state =
+    sensor_hw.initialize(sensor_res, node->get_logger(), node->get_node_clock_interface());
+  ASSERT_EQ(lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED, state.id());
+  ASSERT_EQ(hardware_interface::lifecycle_state_names::UNCONFIGURED, state.label());
+
+  auto state_interfaces = sensor_hw.export_state_interfaces();
+  ASSERT_EQ(2u, state_interfaces.size());
+
+  std::cout << "test 1" << std::endl;
```
Comment (saikishor) [inline hardware_interface/test/test_component_interfaces.cpp:1055]:
```suggestion
```

File: hardware_interface/test/test_component_interfaces.cpp [hunk 3]
```diff
@@ -948,29 +1007,86 @@ TEST(TestComponentInterfaces, dummy_sensor_default)
...
+  auto [contains_sens1_vol, si_sens1_vol] =
+    test_components::vector_contains(state_interfaces, "sens1/voltage");
+  ASSERT_TRUE(contains_sens1_vol);
+  EXPECT_EQ("sens1/voltage", state_interfaces[si_sens1_vol]->get_name());
+  EXPECT_EQ("voltage", state_interfaces[si_sens1_vol]->get_interface_name());
+  EXPECT_EQ("sens1", state_interfaces[si_sens1_vol]->get_prefix_name());
+  EXPECT_TRUE(std::isnan(state_interfaces[si_sens1_vol]->get_value()));
+  std::cout << "test 2" << std::endl;
```
Comment (saikishor) [inline hardware_interface/test/test_component_interfaces.cpp:1063]:
```suggestion
```

File: hardware_interface/test/test_component_interfaces.cpp [hunk 4]
```diff
@@ -948,29 +1007,86 @@ TEST(TestComponentInterfaces, dummy_sensor_default)
...
+
+  auto [contains_joint1_pos, si_joint1_pos] =
+    test_components::vector_contains(state_interfaces, "joint1/position");
+  ASSERT_TRUE(contains_joint1_pos);
+  EXPECT_EQ("joint1/position", state_interfaces[si_joint1_pos]->get_name());
+  EXPECT_EQ("position", state_interfaces[si_joint1_pos]->get_interface_name());
+  EXPECT_EQ("joint1", state_interfaces[si_joint1_pos]->get_prefix_name());
+  EXPECT_TRUE(std::isnan(state_interfaces[si_joint1_pos]->get_value()));
+
+  std::cout << "test 3" << std::endl;
```
Comment (saikishor) [inline hardware_interface/test/test_component_interfaces.cpp:1073]:
```suggestion
```

**Review by saikishor — APPROVED** [review]
Looks good in general.

Maybe we don't need these debug prints

**Review by saikishor — APPROVED** [review]

**Review by bmagyar — APPROVED** [review]

**Pull request merged**